src/api.ts

```diff
@@ -0,0 +1,637 @@
+    ];
+
+    this.method = props.resource.addMethod(props.httpMethod, apiGwIntegration, {
+      methodResponses,
+    });
+  }
+}
+
+export class APIGatewayVTL extends VTL {
+  public integration: IntegrationImpl<AnyFunction> | undefined;
+  constructor(
+    readonly role: aws_iam.IRole,
+    readonly location: "request" | "response",
+    ...statements: string[]
+  ) {
+    super(...statements);
+  }
+
+  protected integrate(
+    target: IntegrationImpl<AnyFunction>,
+    call: CallExpr
+  ): string {
+    if (this.location === "response") {
+      throw new Error(
+        `Cannot call an integration from within a API Gateway Response Template`
+      );
+    }
+    if (target.apiGWVtl) {
+      // ew, mutation
+      // TODO: refactor to pure functions
+      this.integration = target;
+      return target.apiGWVtl.renderRequest(call, this);
+    } else {
+      throw new Error(
+        `Resource type ${target.kind} does not support API Gateway Integrations`
+      );
+    }
+  }
+
+  public eval(node?: Expr, returnVar?: string): string;
+  public eval(node: Stmt, returnVar?: string): void;
+  public eval(node?: Expr | Stmt, returnVar?: string): string | void {
+    if (isReturnStmt(node)) {
+      return this.add(this.json(this.eval(node.expr)));
+    }
+    return super.eval(node as any, returnVar);
+  }
+
+  protected dereference(id: Identifier): string {
+    const ref = id.lookup();
+    if (ref?.kind === "VariableStmt") {
+      return `$${id.name}`;
+    } else if (
+      ref?.kind === "ParameterDecl" &&
+      ref.parent?.kind === "FunctionDecl"
+    ) {
+      const paramIndex = ref.parent.parameters.indexOf(ref);
+      if (paramIndex === 0) {
+        return `$inputRoot`;
+      } else if (paramIndex === 1) {
+        return "$context";
+      } else {
+        throw new Error(`unknown argument`);
+      }
+    }
+    if (id.name.startsWith("$")) {
+      return id.name;
+    } else {
+      return `$${id.name}`;
+    }
+  }
+
+  public json(reference: string): string {
+    return this.jsonStage(reference, 0);
+  }
+
+  private jsonStage(varName: string, level: number): string {
+    if (level === 3) {
+      return "#stop";
+    }
+
+    const itemVarName = this.newLocalVarName();
+    return `#if(${varName}.class.name === 'java.lang.String')
+"${varName}"
+#elseif(${varName}.class.name === 'java.lang.Integer')
+${varName}
+#elseif(${varName}.class.name === 'java.lang.Double')
+${varName}
+#elseif(${varName}.class.name === 'java.lang.Boolean')
+${varName}
+#elseif(${varName}.class.name === 'java.lang.LinkedHashMap')
+{
+#foreach(${itemVarName} in ${varName}.keySet())
+"${itemVarName}": ${this.jsonStage(
+      itemVarName,
+      level + 1
+    )}#if($foreach.hasNext),#end
+#end
+}
+#elseif(${varName}.class.name === 'java.util.ArrayList')
+[
+#foreach(${itemVarName} in ${varName})
+${this.jsonStage(itemVarName, level + 1)}#if($foreach.hasNext),#end
+#end
+]`.replace(/\n/g, `${new Array(level * 2).join(" ")}\n`);
+  }
+}
```

Comment:
This is currently broken and also likely fragile. It generates a explicit recursive loop up to n-depth to render JSON. It quickly explodes in size, but is the only way to serialize arbitrary JSON without `$util.toJson`. Should also look into the solution that would use a go-to.

Reply:
Depth should also be configurable. And it should 500 the request.

Reply:
I abandoned this pattern. We don't support serializing arbitrary JSON objects at this time.